### PR TITLE
Enable Set-Location -LiteralPath to work with folders named "-" and "+"

### LINKED ISSUE #8086: Set-Location -LiteralPath does not treat "-" and "+" as literals.
From great feature, I would say Navigation is impossible to folder named `-` and `+`

image

```Set-Location -LiteralPath -``` as well

Environment data
----------------

```

Name                           Value
----                           -----
PSVersion                      6.2.0-preview.1
PSEdition                      Core
GitCommitId                    6.2.0-preview.1
OS                             Microsoft Windows 10.0.18262 
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
## PR Summary

https://github.com/PowerShell/PowerShell/pull/7206 added the ability to use `cd -` and `cd +` to navigate through location history.  However, it prevented navigating to folders literally named "-" and "+".  The internal `setlocation()` api doesn't differentiate literal and non-literal paths so fix is to overload it with a flag to indicate path is literal so it doesn't process "-" and "+" as shortcuts.

Fix https://github.com/PowerShell/PowerShell/issues/8086

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Review by bergmeister — APPROVED** [review]
LGTM, thanks for quickly taking on this task.

File: src/System.Management.Automation/engine/SessionStateLocationAPIs.cs [hunk 1]
```diff
@@ -194,6 +194,47 @@ internal PathInfo SetLocation(string path)
         /// If the <paramref name="path"/> could not be resolved.
         /// </exception>
         internal PathInfo SetLocation(string path, CmdletProviderContext context)
+        {
+            return SetLocation(path, context, literalPath: false);
+        }
+
+        /// <summary>
+        /// Changes the current working directory to the path specified
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/SessionStateLocationAPIs.cs]: Please add final dot.

File: src/System.Management.Automation/engine/SessionStateLocationAPIs.cs [hunk 2]
```diff
@@ -208,14 +249,14 @@ internal PathInfo SetLocation(string path, CmdletProviderContext context)
 
             switch (originalPath)
             {
-                case string originalPathSwitch when originalPathSwitch.Equals("-", StringComparison.OrdinalIgnoreCase):
+                case string originalPathSwitch when !literalPath && originalPathSwitch.Equals("-", StringComparison.OrdinalIgnoreCase):
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/SessionStateLocationAPIs.cs]:
```suggestion
                case string originalPathSwitch when !literalPath && originalPathSwitch.Equals("-", StringComparison.Ordinal):
```

File: src/System.Management.Automation/engine/SessionStateLocationAPIs.cs [hunk 3]
```diff
@@ -208,14 +249,14 @@ internal PathInfo SetLocation(string path, CmdletProviderContext context)
...
                     if (_setLocationHistory.UndoCount <= 0)
                     {
                         throw new InvalidOperationException(SessionStateStrings.LocationUndoStackIsEmpty);
                     }
                     path = _setLocationHistory.Undo(this.CurrentLocation).Path;
                     break;
-                case string originalPathSwitch when originalPathSwitch.Equals("+", StringComparison.OrdinalIgnoreCase):
+                case string originalPathSwitch when !literalPath && originalPathSwitch.Equals("+", StringComparison.OrdinalIgnoreCase):
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/SessionStateLocationAPIs.cs]:
```suggestion
                case string originalPathSwitch when !literalPath && originalPathSwitch.Equals("+", StringComparison.Ordinal):
```

File: src/System.Management.Automation/engine/SessionStateLocationAPIs.cs [hunk 4]
```diff
@@ -194,6 +194,47 @@ internal PathInfo SetLocation(string path)
...
+        /// If <paramref name="path"/> refers to a drive that does not exist.
+        /// </exception>
+        /// <exception cref="ProviderInvocationException">
+        /// If the provider associated with <paramref name="path"/> threw an
+        /// exception.
+        /// </exception>
+        /// <exception cref="ItemNotFoundException">
+        /// If the <paramref name="path"/> could not be resolved.
+        /// </exception>
+        internal PathInfo SetLocation(string path, CmdletProviderContext context, bool literalPath)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/SessionStateLocationAPIs.cs]: Should it be private now?

Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/SessionStateLocationAPIs.cs]: No, it's needed to be called from PathInterfaces

**Review by iSazonov — APPROVED** [review]

**Review by anmenaga — APPROVED** [review]
It would be good to have a `[feature]` test pass for this.

**Pull request merged**